src/codecs/csl/index.ts

```diff
@@ -0,0 +1,228 @@
+/**
+ * Codec for CSL JSON
+ *
+ * Also acts a base codec for other bibliographic formats
+ * e.g. BibTeX. Based on https://citation.js.org/
+ */
+
+import * as stencila from '@stencila/schema'
+import { isA, isCreativeWork } from '@stencila/schema/dist/util'
+// @ts-ignore
+import Cite from 'citation-js'
+import Csl from 'csl-json'
+import { Encode, EncodeOptions, load } from '../..'
+import * as vfile from '../../util/vfile'
+import { errorNodeType, warnLossIfAny, warnLoss } from '../../log'
+
+export const mediaTypes = ['application/vnd.citationstyles.csl+json']
+
+export const extNames = ['csl']
+
+/**
+ * Parse CSL-JSON or other bibliographic format to a `Node`
+ */
+export async function decode(
+  file: vfile.VFile,
+  format: string = '@csl/object'
+): Promise<stencila.Node> {
+  const content: string = await vfile.dump(file)
+
+  let csls
+  try {
+    csls = await Cite.inputAsync(content, { forceType: format })
+  } catch (error) {
+    throw new Error(
+      `Error when parsing content of format ${format}: ${error.message}`
+    )
+  }
+
+  // TODO: work out what to return when more than one work e.g. a bibtex file
+  const csl = csls[0]
+
+  return decodeCsl(csl)
+}
+
+/**
+ * Encode a `Node` to CSL-JSON or other bibliographic format
+ *
+ * See https://citation.js.org/api/tutorial-output_options.html
+ * for formats and other options which could be used.
+ */
+export const encode: Encode = async (
+  node: stencila.Node,
+  options: EncodeOptions = {}
+): Promise<vfile.VFile> => {
+  const { format = 'json' } = options
+
+  let content = ''
+  if (isCreativeWork(node)) {
+    const csl = encodeCsl(node)
+    const cite = new Cite([csl])
+    if (format === 'json') {
+      const { _graph, ...rest } = cite.data[0]
+      content = JSON.stringify(rest, null, 2)
+    } else {
+      content = cite.format(format)
+    }
+  } else {
+    errorNodeType('csl', 'encode', 'CreativeWork', node)
+  }
+
+  return vfile.load(content)
+}
+
+/**
+ * Decode `Csl.Data` to a `CreativeWork` or derived type.
+ */
+export async function decodeCsl(
+  csl: Csl.Data
+): Promise<stencila.CreativeWork | stencila.Article> {
+  const {
+    type,
+    id,
+
+    author = [],
+    title = '',
+
+    issued,
+
+    'container-title': containerTitle,
+    volume,
+    issue,
+    page,
+
+    // @ts-ignore // Citation.js uses this hidden property
+    _graph,
+
+    ...lost
+  } = csl
```

Comment:
Use object destructing to concisely (a) set default values (instead of `const author = csl.author !== undefined ? csl.author : []`), and (b) get `lost` which is an object with all the properties that we don't explicitly handle below.